python/graphstorm/wholegraph/wholegraph.py

```diff
@@ -457,16 +477,100 @@ def load_from_file(
             The prefix of the file.
         num_files : int
             The number of files to load.
-
+        wg_optimizer : WholeMemoryOptimizer or None
+            The WholeGraph sparse optimizer to be attached to the WholeGraph embedding.
         Returns
         -------
         None
         """
+
+        if wg_optimizer is not None:
+            assert self._use_wg_optimizer, \
+                "Please create WholeGraphDistTensor tensor with use_wg_optimizer=True."
+            # attach to an optimizer
+            self.attach_wg_optimizer(wg_optimizer)
+        else:
+            if self.use_wg_optimizer:
+                assert self.optimizer is not None, \
+                    "Need either self.optimizer or wg_optimizer to be available."
+                assert self._module is not None and self._tensor is not None, \
+                    "Please create WholeGraphDistTensor tensor with attach_wg_optimizer()."
+            else:
+                assert self.optimizer is None and self._module is None, \
+                    "For regular embeddings (not trainable), self.optimizer should be None."
+        # replace the existing _tensor by loading from file
         file_prefix = os.path.join(path, file_prefix)
-        self._module.wm_embedding.get_embedding_tensor().from_file_prefix(
+        self._tensor.get_embedding_tensor().from_file_prefix(
             file_prefix, part_count=num_files
         )
 
+    def __setitem__(self, idx: th.Tensor, val: th.Tensor):
+        """
+        Set the embeddings for the specified node indices.
+        This call must be called by all processes.
+
+        Parameters
+        ----------
+        idx : torch.Tensor
+            Index of the embeddings to collect.
+        val : torch.Tensor
+            The requested node embeddings.
+        """
+        assert self._tensor is not None, "Please create WholeGraph tensor first."
+        idx = idx.cuda()
+        val = val.cuda()
+
+        if val.dtype != self.dtype:
+            val = val.to(self.dtype)
+        self._tensor.get_embedding_tensor().scatter(val, idx)
+
+    def __getitem__(self, idx: th.Tensor) -> th.Tensor:
+        """
+        Get the embeddings for the specified node indices (remotely).
+        This call must be called by all processes.
+
+        Parameters
+        ----------
+        idx : torch.Tensor
+            Index of the embeddings to collect.
+        Returns
+        -------
+        torch.Tensor
+            The requested node embeddings.
+        """
+        assert self._tensor is not None, "Please create WholeGraph tensor first."
+        idx = idx.cuda()
+        output_tensor = self._tensor.gather(idx)  # output_tensor is on cuda by default
+        return output_tensor
+
+    def get_local_tensor(self):
+        """ Get the local tensor and offset of the embedding tensor. """
```

Comment:
What is the type of local_tensor. Can you add a docstring for Return values?

Reply:
Added docstring for Returns